src/test/compile-fail/privacy-struct-ctor.rs

```diff
@@ -0,0 +1,42 @@
+// Copyright 2016 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// aux-build:privacy-struct-ctor.rs
+
+#![feature(pub_restricted)]
+
+extern crate privacy_struct_ctor as xcrate;
+
+mod m {
+    pub struct S(u8);
+
+    pub mod n {
+        pub(m) struct Z(pub(m::n) u8);
+    }
+
+    use m::n::Z; // OK, only the type is imported
+
+    fn f() {
+        n::Z; //~ ERROR tuple struct `Z` is private
+        Z; //~ ERROR expected value, found struct `Z`
+    }
+}
+
+use m::S; // OK, only the type is imported
+
+fn main() {
+    m::S; //~ ERROR tuple struct `S` is private
+    S; //~ ERROR expected value, found struct `S`
```

Comment:
@petrochenkov 
> Is there any way to map struct's DefId into (ctor's DefId + ctor's ty::Visibility) on demand without building tables in advance?

We could simplify a little by mapping the struct's `DefId` to the ctor's `&'a NameBinding<'a>`, but we'd still be building an extra table.

To avoid that, we could add a variant
```rust
NameBindingKind::TupleStruct(Def, &'a NameBinding<'a> /* ctor */)
```
and use that for tuple struct type bindings instead of `NameBindingKind::Def`.
`NameBindingKind` is well encapsulated, so that shouldn't be too big of a change, but it still might not be worth it.